src/atc/sql/schema.py

```diff
@@ -1,16 +1,315 @@
+                    f"name:type, found after {repr(field_name)}: {_flatten(it)}"
+                )
+            next(it)  # consume the colon
+
+            # the type can also be a struct again. Therefore, recurse here.
+            field_type = _get_data_type(it)
+
+            struct.add(field_name, field_type)
+            c = next(it)
+            if c == ",":
+                # get next field
+                continue
+            elif c == ">":
+                break
+            else:
+                raise SchemaExtractionError(f'invalid char "{c}" after schema fields')
+        return struct
+    elif type_name == "ARRAY":
+        # we need to parse the array element type
+        c = next(it)
+
+        if c != "<":
+            raise SchemaExtractionError(
+                f"after ARRAY must follow a < > block. found: {_flatten(it)}"
+            )
+        element_type = _get_data_type(it)
+        c = next(it)
+        if c != ">":
+            raise SchemaExtractionError(
+                f"after ARRAY<{element_type} must follow a >. found: {c}"
+            )
+        return t.ArrayType(elementType=element_type)
+    elif type_name == "MAP":
+        # we need to parse the array element type
+        c = next(it)
+
+        if c != "<":
+            raise SchemaExtractionError(
+                f"after MAP must follow a < > block. found: {_flatten(it)}"
+            )
+        key_type = _get_data_type(it)
+
+        if it.peek() != ",":
+            raise SchemaExtractionError(
+                f"after MAP<KeyType must follow a ,. found: {_flatten(it)}"
+            )
+        next(it)
+
+        value_type = _get_data_type(it)
+
+        if it.peek() != ">":
+            raise SchemaExtractionError(
+                f"after MAP<KeyType,ValueType must follow a >. found: {_flatten(it)}"
+            )
+        next(it)
+
+        return t.MapType(key_type, value_type)
+
+    else:
+        raise NotImplementedError(f"Data type not implementd: {type_name}")
+
+
+def _get_ascii_token(it: peekable) -> str:
+    name = []
+
+    while it.peek(".") in string.ascii_letters:
+        name.append(next(it))
+
+    name = "".join(name).strip()
+
+    if not name:
+        raise SchemaExtractionError(f"unable to get a token: '{_flatten(it)}'")
+    return name
+
+
+def _get_identifier(it: peekable) -> str:
+    escaped = False
+    name = []
+    first = next(it)
+
+    # let's get this started.
+    # I'm using https://spark.apache.org/docs/latest/sql-ref-identifier.html
+    if first == "`":
+        escaped = True
+    elif first in string.ascii_letters:
+        name.append(first)
+    else:
+        SchemaExtractionError("Invalid start of identifier sequence.")
+
+    non_escaped_characters = string.ascii_letters + string.digits + "_"
+    non_whitespace = non_escaped_characters + string.punctuation
+
+    while True:
+        char = it.peek()
+
+        if char == "`":
+            escaped = not escaped
```

Comment:
what happens here with escaped = not escaped?

Reply:
it inverts the boolean. But no longer relevant.